# Data Mining – Assignment 1
#### Department of Management Science and Technology 
#### Master in Business Analytics 
#### Athens University of Economics and Business
---
>Christos Kallaras, p2822009 <br />

In this assignment we will implement a simple workflow that will assess the similarity
between supermarket customers. Then we will use these results to predict the rating of a customer. 

In [ ]:
import pandas as pd
import numpy as np
import matplotlib
from tabulate import tabulate

Note : In order to run the paper correctly enter the path of the file in this literal.

In [ ]:
path = 'C:/Users/Christos Kallaras/Desktop/Master/1st Year/3rd Trimester/Mining Big Datasets/1st Assignment/'

In [ ]:
#Read the file
groceries = pd.read_csv(path + 'groceries.csv',sep=";")
print(groceries.shape)
groceries.head(20)

(10000, 10)


Customer_ID Age     Sex Marital_Status  Education Income Customer_Rating  \
0             1  75    male        married    primary  20000       very_good   
1             2  61  female         single  secondary  28000            good   
2             3  32    male         single  secondary  34000       very_good   
3             4  62    male        married    primary  31000       very_good   
4             5  66  female        married  secondary  19000            good   
5             6  55  female         single  secondary  35000       very_good   
6             7  23  female        married   tertiary  21000            good   
7             8  26  female         single  secondary  30000            good   
8             9  29  female        married  secondary  32000       very_good   
9            10  57  female        married  secondary  26000            good   
10           11  55  female        married   tertiary  30000       very_good   
11           12  57  female        married   tertiary  14000       very_good   
12           13  28    male       divorced   tertiary  30000            fair   
13           14  40    male        married    primary  13000            fair   
14           15  53  female       divorced   tertiary   9000       very_good   
15           16      female        married   tertiary  25000            good   
16           17  68  female        married  secondary                   good   
17           18  32    male        married    primary  16000       very_good   
18           19  33    male         single    primary  18000       very_good   
19           20  23    male        married   tertiary  23000            fair   

    Persons_in_Household     Occupation  \
0                      3        retired   
1                      1      housemaid   
2                      1    blue-collar   
3                      3    blue-collar   
4                      3        retired   
5                      1     unemployed   
6                      3      housemaid   
7                      2    blue-collar   
8                      3    blue-collar   
9                      3   entrepreneur   
10                     2   entrepreneur   
11                     3    blue-collar   
12                     1  self-employed   
13                     3   entrepreneur   
14                     1   entrepreneur   
15                     4   entrepreneur   
16                     3        retired   
17                     3   entrepreneur   
18                     1    blue-collar   
19                     3   entrepreneur   

                                            Groceries  
0   citrus fruit,semi-finished bread,margarine,rea...  
1                        tropical fruit,yogurt,coffee  
2                                          whole milk  
3          pip fruit,yogurt,cream cheese,meat spreads  
4   other vegetables,whole milk,condensed milk,lon...  
5      whole milk,butter,yogurt,rice,abrasive cleaner  
6                                          rolls/buns  
7   other vegetables,UHT-milk,rolls/buns,bottled b...  
8                                       potted plants  
9                                  whole milk,cereals  
10  tropical fruit,other vegetables,white bread,bo...  
11  citrus fruit,tropical fruit,whole milk,butter,...  
12                                               beef  
13                        frankfurter,rolls/buns,soda  
14                             chicken,tropical fruit  
15      butter,sugar,fruit/vegetable juice,newspapers  
16                              fruit/vegetable juice  
17                          packaged fruit/vegetables  
18                                          chocolate  
19                                      specialty bar

In [ ]:
groceries.dtypes

Customer_ID              int64
Age                     object
Sex                     object
Marital_Status          object
Education               object
Income                  object
Customer_Rating         object
Persons_in_Household     int64
Occupation              object
Groceries               object
dtype: object

The missing values are whitespace ,not null so in order to transform them we replace them with 0.

In [ ]:
groceries['Age'] = np.where((groceries.Age == ' '),'0',groceries.Age)
groceries['Income'] = np.where((groceries.Income == ' '),'0',groceries.Income)
groceries.head(20)

Customer_ID Age     Sex Marital_Status  Education Income Customer_Rating  \
0             1  75    male        married    primary  20000       very_good   
1             2  61  female         single  secondary  28000            good   
2             3  32    male         single  secondary  34000       very_good   
3             4  62    male        married    primary  31000       very_good   
4             5  66  female        married  secondary  19000            good   
5             6  55  female         single  secondary  35000       very_good   
6             7  23  female        married   tertiary  21000            good   
7             8  26  female         single  secondary  30000            good   
8             9  29  female        married  secondary  32000       very_good   
9            10  57  female        married  secondary  26000            good   
10           11  55  female        married   tertiary  30000       very_good   
11           12  57  female        married   tertiary  14000       very_good   
12           13  28    male       divorced   tertiary  30000            fair   
13           14  40    male        married    primary  13000            fair   
14           15  53  female       divorced   tertiary   9000       very_good   
15           16   0  female        married   tertiary  25000            good   
16           17  68  female        married  secondary      0            good   
17           18  32    male        married    primary  16000       very_good   
18           19  33    male         single    primary  18000       very_good   
19           20  23    male        married   tertiary  23000            fair   

    Persons_in_Household     Occupation  \
0                      3        retired   
1                      1      housemaid   
2                      1    blue-collar   
3                      3    blue-collar   
4                      3        retired   
5                      1     unemployed   
6                      3      housemaid   
7                      2    blue-collar   
8                      3    blue-collar   
9                      3   entrepreneur   
10                     2   entrepreneur   
11                     3    blue-collar   
12                     1  self-employed   
13                     3   entrepreneur   
14                     1   entrepreneur   
15                     4   entrepreneur   
16                     3        retired   
17                     3   entrepreneur   
18                     1    blue-collar   
19                     3   entrepreneur   

                                            Groceries  
0   citrus fruit,semi-finished bread,margarine,rea...  
1                        tropical fruit,yogurt,coffee  
2                                          whole milk  
3          pip fruit,yogurt,cream cheese,meat spreads  
4   other vegetables,whole milk,condensed milk,lon...  
5      whole milk,butter,yogurt,rice,abrasive cleaner  
6                                          rolls/buns  
7   other vegetables,UHT-milk,rolls/buns,bottled b...  
8                                       potted plants  
9                                  whole milk,cereals  
10  tropical fruit,other vegetables,white bread,bo...  
11  citrus fruit,tropical fruit,whole milk,butter,...  
12                                               beef  
13                        frankfurter,rolls/buns,soda  
14                             chicken,tropical fruit  
15      butter,sugar,fruit/vegetable juice,newspapers  
16                              fruit/vegetable juice  
17                          packaged fruit/vegetables  
18                                          chocolate  
19                                      specialty bar

To calculate the mean we need the numerical values to be numeric so we tranform them.

In [ ]:
groceries['Age'] = groceries['Age'].astype(str).astype(int)
groceries['Income'] = groceries['Income'].astype(str).astype(int)
groceries['Persons_in_Household'] = groceries['Persons_in_Household'].astype(str).astype(int)
print(groceries.dtypes)

Customer_ID              int64
Age                      int32
Sex                     object
Marital_Status          object
Education               object
Income                   int32
Customer_Rating         object
Persons_in_Household     int32
Occupation              object
Groceries               object
dtype: object


Now we can calculate the average, by replacing 0 with the mean.

In [ ]:
groceries['Age'] = np.where((groceries.Age == 0),groceries.Age.mean(),groceries.Age)
groceries['Income'] = np.where((groceries.Income == 0),groceries.Income.mean(),groceries.Income)
groceries.head(20)

Customer_ID      Age     Sex Marital_Status  Education   Income  \
0             1  75.0000    male        married    primary  20000.0   
1             2  61.0000  female         single  secondary  28000.0   
2             3  32.0000    male         single  secondary  34000.0   
3             4  62.0000    male        married    primary  31000.0   
4             5  66.0000  female        married  secondary  19000.0   
5             6  55.0000  female         single  secondary  35000.0   
6             7  23.0000  female        married   tertiary  21000.0   
7             8  26.0000  female         single  secondary  30000.0   
8             9  29.0000  female        married  secondary  32000.0   
9            10  57.0000  female        married  secondary  26000.0   
10           11  55.0000  female        married   tertiary  30000.0   
11           12  57.0000  female        married   tertiary  14000.0   
12           13  28.0000    male       divorced   tertiary  30000.0   
13           14  40.0000    male        married    primary  13000.0   
14           15  53.0000  female       divorced   tertiary   9000.0   
15           16  50.8568  female        married   tertiary  25000.0   
16           17  68.0000  female        married  secondary  28604.2   
17           18  32.0000    male        married    primary  16000.0   
18           19  33.0000    male         single    primary  18000.0   
19           20  23.0000    male        married   tertiary  23000.0   

   Customer_Rating  Persons_in_Household     Occupation  \
0        very_good                     3        retired   
1             good                     1      housemaid   
2        very_good                     1    blue-collar   
3        very_good                     3    blue-collar   
4             good                     3        retired   
5        very_good                     1     unemployed   
6             good                     3      housemaid   
7             good                     2    blue-collar   
8        very_good                     3    blue-collar   
9             good                     3   entrepreneur   
10       very_good                     2   entrepreneur   
11       very_good                     3    blue-collar   
12            fair                     1  self-employed   
13            fair                     3   entrepreneur   
14       very_good                     1   entrepreneur   
15            good                     4   entrepreneur   
16            good                     3        retired   
17       very_good                     3   entrepreneur   
18       very_good                     1    blue-collar   
19            fair                     3   entrepreneur   

                                            Groceries  
0   citrus fruit,semi-finished bread,margarine,rea...  
1                        tropical fruit,yogurt,coffee  
2                                          whole milk  
3          pip fruit,yogurt,cream cheese,meat spreads  
4   other vegetables,whole milk,condensed milk,lon...  
5      whole milk,butter,yogurt,rice,abrasive cleaner  
6                                          rolls/buns  
7   other vegetables,UHT-milk,rolls/buns,bottled b...  
8                                       potted plants  
9                                  whole milk,cereals  
10  tropical fruit,other vegetables,white bread,bo...  
11  citrus fruit,tropical fruit,whole milk,butter,...  
12                                               beef  
13                        frankfurter,rolls/buns,soda  
14                             chicken,tropical fruit  
15      butter,sugar,fruit/vegetable juice,newspapers  
16                              fruit/vegetable juice  
17                          packaged fruit/vegetables  
18                                          chocolate  
19                                      specialty bar

Finally we transform them to integers in order to keep only the integer part as instructed by the assingment.

In [ ]:
groceries['Age'] = groceries['Age'].astype(int)
groceries['Income'] = groceries['Income'].astype(int)
groceries['Persons_in_Household'] = groceries['Persons_in_Household'].astype(int)
groceries.head(20)

Customer_ID  Age     Sex Marital_Status  Education  Income  \
0             1   75    male        married    primary   20000   
1             2   61  female         single  secondary   28000   
2             3   32    male         single  secondary   34000   
3             4   62    male        married    primary   31000   
4             5   66  female        married  secondary   19000   
5             6   55  female         single  secondary   35000   
6             7   23  female        married   tertiary   21000   
7             8   26  female         single  secondary   30000   
8             9   29  female        married  secondary   32000   
9            10   57  female        married  secondary   26000   
10           11   55  female        married   tertiary   30000   
11           12   57  female        married   tertiary   14000   
12           13   28    male       divorced   tertiary   30000   
13           14   40    male        married    primary   13000   
14           15   53  female       divorced   tertiary    9000   
15           16   50  female        married   tertiary   25000   
16           17   68  female        married  secondary   28604   
17           18   32    male        married    primary   16000   
18           19   33    male         single    primary   18000   
19           20   23    male        married   tertiary   23000   

   Customer_Rating  Persons_in_Household     Occupation  \
0        very_good                     3        retired   
1             good                     1      housemaid   
2        very_good                     1    blue-collar   
3        very_good                     3    blue-collar   
4             good                     3        retired   
5        very_good                     1     unemployed   
6             good                     3      housemaid   
7             good                     2    blue-collar   
8        very_good                     3    blue-collar   
9             good                     3   entrepreneur   
10       very_good                     2   entrepreneur   
11       very_good                     3    blue-collar   
12            fair                     1  self-employed   
13            fair                     3   entrepreneur   
14       very_good                     1   entrepreneur   
15            good                     4   entrepreneur   
16            good                     3        retired   
17       very_good                     3   entrepreneur   
18       very_good                     1    blue-collar   
19            fair                     3   entrepreneur   

                                            Groceries  
0   citrus fruit,semi-finished bread,margarine,rea...  
1                        tropical fruit,yogurt,coffee  
2                                          whole milk  
3          pip fruit,yogurt,cream cheese,meat spreads  
4   other vegetables,whole milk,condensed milk,lon...  
5      whole milk,butter,yogurt,rice,abrasive cleaner  
6                                          rolls/buns  
7   other vegetables,UHT-milk,rolls/buns,bottled b...  
8                                       potted plants  
9                                  whole milk,cereals  
10  tropical fruit,other vegetables,white bread,bo...  
11  citrus fruit,tropical fruit,whole milk,butter,...  
12                                               beef  
13                        frankfurter,rolls/buns,soda  
14                             chicken,tropical fruit  
15      butter,sugar,fruit/vegetable juice,newspapers  
16                              fruit/vegetable juice  
17                          packaged fruit/vegetables  
18                                          chocolate  
19                                      specialty bar

As we can see customer with id 15 went from not having a value to Age, to now having 50. Similar example is customer 16 and his income.

##  Compute data (dis-)similarity

To assess the similarity between the customers we first have to distinguish the type of each attribute :
- Age is numerical
- Sex is categorical
- Marital_Status is categorical
- Education is ordinal
- Income is numerical
- Customer_rating is ordinal
- Persons_in_Household is ordinal
- Occupation is categorical
- Groceries is set

For the ordinal attributes we need to rank them in order to use them for our calculations.

In [ ]:
#check values for ordinal variables

print(groceries["Education"].value_counts())
print(groceries["Customer_Rating"].value_counts())
print(groceries["Persons_in_Household"].value_counts())

tertiary     3384
secondary    3358
primary      3258
Name: Education, dtype: int64
good         3121
very_good    2374
fair         2246
excellent    1135
poor         1124
Name: Customer_Rating, dtype: int64
3    3105
1    2907
2    2317
4    1330
5     313
6      28
Name: Persons_in_Household, dtype: int64


In [ ]:
#convert ordinal variable values to numbers

fix_education = {"Education":     {"primary": 1, "secondary": 2,"tertiary": 3}}
fix_cust_rating = {"Customer_Rating":     {"poor": 1, "fair": 2, "good": 3,"very_good": 4,"excellent": 5}}
groceries.replace(fix_education, inplace=True)
groceries.replace(fix_cust_rating, inplace=True)
groceries.head() 

Customer_ID  Age     Sex Marital_Status  Education  Income  \
0            1   75    male        married          1   20000   
1            2   61  female         single          2   28000   
2            3   32    male         single          2   34000   
3            4   62    male        married          1   31000   
4            5   66  female        married          2   19000   

   Customer_Rating  Persons_in_Household   Occupation  \
0                4                     3      retired   
1                3                     1    housemaid   
2                4                     1  blue-collar   
3                4                     3  blue-collar   
4                3                     3      retired   

                                           Groceries  
0  citrus fruit,semi-finished bread,margarine,rea...  
1                       tropical fruit,yogurt,coffee  
2                                         whole milk  
3         pip fruit,yogurt,cream cheese,meat spreads  
4  other vegetables,whole milk,condensed milk,lon...

As for attribute Groceries which is a set, we transform its element into lists.

In [ ]:
groceries['Groceries']  = groceries['Groceries'].str.split(',')
groceries.head(5)

Customer_ID  Age     Sex Marital_Status  Education  Income  \
0            1   75    male        married          1   20000   
1            2   61  female         single          2   28000   
2            3   32    male         single          2   34000   
3            4   62    male        married          1   31000   
4            5   66  female        married          2   19000   

   Customer_Rating  Persons_in_Household   Occupation  \
0                4                     3      retired   
1                3                     1    housemaid   
2                4                     1  blue-collar   
3                4                     3  blue-collar   
4                3                     3      retired   

                                           Groceries  
0  [citrus fruit, semi-finished bread, margarine,...  
1                   [tropical fruit, yogurt, coffee]  
2                                       [whole milk]  
3    [pip fruit, yogurt, cream cheese, meat spreads]  
4  [other vegetables, whole milk, condensed milk,...

We check if there are any more null values.

In [ ]:
null_values = groceries[groceries.isnull().any(axis=1) == True].index.tolist()
null_values

[]

We do not have null values so we continue.

We created a function that calculates the similarities between 2 ids. Since we need it also for predicting the rating of a customer ,in which case we will not be having the attribute Customer_Rate, we apply a check; if the daraframe has 10 columns ,meaning that we have all the columns , then calculate the similarities based also on Customer_Rating, else remove Customer_Rating from the calculation. That way we can use the same function in predicting the rating.

The calculation happens as follows 
- for numerical attributes     : **d(a,b) = |a - b| / (max_value - min_value)**
- for ordinal attributes       :  **d(a,b) = |rank(a) - rank(b)| / (max_value - min_value)**
- for categorical attributes   :  **d(a,b) = 0 if a=b , 1 if a<>b**
- for set                      : calculate jaccard distance **dJaccard(a,b) = 1 - intersection(s1,s2)/union(s1,s2)**

Then it returns the average.<br />
The input for the function is the customer ids for which we want to make the calculation and the dataset.

In [ ]:
#create function that calculates dis for given ids

def calc_dis(id1,id2,groceries):
    
    #Since numbering starts from 0 in the df but customers start at 1 we substract 1 to find the corresponding ids
    #For example if we want to find disimilarities for customer 1 we actualy searching for row with id 0
    id1 = id1 - 1
    id2 = id2 - 1
    
    #numerical
    #age
    Age_max = groceries['Age'].max()
    Age_min = groceries['Age'].min()
    
    dis_age = abs(groceries.iloc[id1].Age-groceries.iloc[id2].Age)/(Age_max-Age_min)
    
    #income
    Income_max = groceries['Income'].max()
    Income_min = groceries['Income'].min()

    dis_income = abs(groceries.iloc[id1].Income-groceries.iloc[id2].Income)/(Income_max-Income_min)
    
    #ordinal
    Education_max = groceries['Education'].max()
    Education_min = groceries['Education'].min()
    
    dis_edu = abs(groceries.iloc[id1].Education-groceries.iloc[id2].Education)/(Education_max-Education_min)
    
    #Our dataset will have the column Customer_Rating only for questions 2 and 3. 
    #For the 4th we will remove it so the columns number will be 9
    if (len(groceries.columns) == 10):
        
        Customer_Rating_max = groceries['Customer_Rating'].max()
        Customer_Rating_min = groceries['Customer_Rating'].min()
    
        dis_cust_rat = abs(groceries.iloc[id1].Customer_Rating-groceries.iloc[id2].Customer_Rating)/(Customer_Rating_max-Customer_Rating_min)
    
    Persons_in_Household_max = groceries['Persons_in_Household'].max()
    Persons_in_Household_min = groceries['Persons_in_Household'].min()
    
    dis_per = abs(groceries.iloc[id1].Persons_in_Household-groceries.iloc[id2].Persons_in_Household)/(Persons_in_Household_max-Persons_in_Household_min)
    
    #categorical
    #sex
    if groceries.iloc[id1].Sex == groceries.iloc[id2].Sex:
        dis_sex = 0
    else:
        dis_sex = 1
    
    #Marital_Status
    if groceries.iloc[id1].Marital_Status == groceries.iloc[id2].Marital_Status:
        dis_mar = 0
    else:
        dis_mar = 1
        
    #Occupation
    if groceries.iloc[id1].Occupation == groceries.iloc[id2].Occupation:
        dis_occ = 0
    else:
        dis_occ = 1    
    
    #Groceries
    intersection = len(list(set(groceries.iloc[id1].Groceries).intersection(groceries.iloc[id2].Groceries)))
    union = (len(groceries.iloc[id1].Groceries) + len(groceries.iloc[id2].Groceries)) - intersection
    dis_groc = 1 - (float(intersection) / union)
    
    
    #return the average
    if (len(groceries.columns) == 10):
        return (dis_age+dis_income+dis_edu+dis_cust_rat+dis_per+dis_sex+dis_mar+dis_occ+dis_groc)/9
    else:
        return (dis_age+dis_income+dis_edu+dis_per+dis_sex+dis_mar+dis_occ+dis_groc)/8

Since our machines could not handle running the function for all customer ids and create the whole dissimilarity matrices, we have 3 examples of calling the function.

For example calculate the disimilarity between customer 1 and 7749, then 4 and 653 and 73 with 2078.

In [ ]:
#example calculate dissimilarity
print(calc_dis(1,7749,groceries))
print(calc_dis(4,653,groceries))
print(calc_dis(73,2078,groceries))

0.1002978502978503
0.5994690494690494
0.30852110852110853


We can se that customer 1 is quite similar to customer 7749 but no so much to customer 653. If the disimilarity score was 1 then they would be completely different.

##  Nearest Neighbor (NN) search

In order to find the 10 most similar customers we created a function that takes as input the customer id for whom we want to make the search and the dataset. Then it calls the previous function and calculates the similarities for all the other customers (except the one for which we perform the search). The results are stored in a dictionary and then sorted in a list by ascending order. We return the first 10, the ones with the smallest score since we want similarities.

In [ ]:
def calc_10nn(id,groceries):
    dis_dic = {}
    for i, row in groceries.iterrows():
        #we exclude the id for which the search is happening
        if i != id :
            #call previous function in order to calculate dissimilarity
            dis_dic[i] = calc_dis(id,i,groceries)
    #sort the dictionary and store it in a list ascending
    dis_list = sorted(dis_dic.items(), key=lambda x: x[1])
    #we need similar so take top 10 since we sorted on ascending order
    sort_dis_list = dis_list[:10]
    #insert column names
    sort_dis_list.insert(0,('id','similarity'))
    return sort_dis_list

For testing purposes we print the 10 NN most similar customers to customer 1.

In [ ]:
print (tabulate(calc_10nn(1,groceries),headers='firstrow',tablefmt='fancy_grid'))

╒══════╤══════════════╕
│   id │   similarity │
╞══════╪══════════════╡
│ 7749 │     0.100298 │
├──────┼──────────────┤
│ 7931 │     0.101036 │
├──────┼──────────────┤
│ 9514 │     0.110464 │
├──────┼──────────────┤
│  628 │     0.112073 │
├──────┼──────────────┤
│ 6918 │     0.114016 │
├──────┼──────────────┤
│ 4230 │     0.119632 │
├──────┼──────────────┤
│ 3148 │     0.121238 │
├──────┼──────────────┤
│ 4647 │     0.121652 │
├──────┼──────────────┤
│ 2105 │     0.122999 │
├──────┼──────────────┤
│ 8050 │     0.124605 │
╘══════╧══════════════╛


We can see that they are exactly the same with the ones given to us by the assingment for customer 1, so we procceed with the customers for which we want to find the 10 NN most similars.

- For customer 73

In [ ]:
print (tabulate(calc_10nn(73,groceries),headers='firstrow',tablefmt='fancy_grid'))

╒══════╤══════════════╕
│   id │   similarity │
╞══════╪══════════════╡
│ 1846 │     0.122537 │
├──────┼──────────────┤
│ 1291 │     0.129863 │
├──────┼──────────────┤
│ 1203 │     0.143512 │
├──────┼──────────────┤
│ 5881 │     0.146775 │
├──────┼──────────────┤
│ 1627 │     0.149068 │
├──────┼──────────────┤
│ 3953 │     0.152512 │
├──────┼──────────────┤
│ 5922 │     0.156533 │
├──────┼──────────────┤
│ 6904 │     0.157854 │
├──────┼──────────────┤
│ 8881 │     0.159311 │
├──────┼──────────────┤
│ 3623 │     0.160425 │
╘══════╧══════════════╛


- For customer 563

In [ ]:
print (tabulate(calc_10nn(563,groceries),headers='firstrow',tablefmt='fancy_grid'))

╒══════╤══════════════╕
│   id │   similarity │
╞══════╪══════════════╡
│ 3634 │    0.0696193 │
├──────┼──────────────┤
│ 6168 │    0.0929811 │
├──────┼──────────────┤
│ 2839 │    0.118842  │
├──────┼──────────────┤
│ 6196 │    0.124657  │
├──────┼──────────────┤
│ 2766 │    0.126289  │
├──────┼──────────────┤
│ 8108 │    0.12647   │
├──────┼──────────────┤
│  559 │    0.129656  │
├──────┼──────────────┤
│ 6929 │    0.131572  │
├──────┼──────────────┤
│ 9578 │    0.13402   │
├──────┼──────────────┤
│ 8270 │    0.13639   │
╘══════╧══════════════╛


- For customer 1603

In [ ]:
print (tabulate(calc_10nn(1603,groceries),headers='firstrow',tablefmt='fancy_grid'))

╒══════╤══════════════╕
│   id │   similarity │
╞══════╪══════════════╡
│ 7345 │     0.131391 │
├──────┼──────────────┤
│ 9260 │     0.139128 │
├──────┼──────────────┤
│ 7335 │     0.140089 │
├──────┼──────────────┤
│  568 │     0.153047 │
├──────┼──────────────┤
│  109 │     0.153587 │
├──────┼──────────────┤
│ 4814 │     0.154066 │
├──────┼──────────────┤
│ 6751 │     0.160826 │
├──────┼──────────────┤
│ 4628 │     0.162134 │
├──────┼──────────────┤
│  168 │     0.164025 │
├──────┼──────────────┤
│ 6841 │     0.167444 │
╘══════╧══════════════╛


- For customer 2200

In [ ]:
print (tabulate(calc_10nn(2200,groceries),headers='firstrow',tablefmt='fancy_grid'))

╒══════╤══════════════╕
│   id │   similarity │
╞══════╪══════════════╡
│  403 │     0.144147 │
├──────┼──────────────┤
│ 7497 │     0.157459 │
├──────┼──────────────┤
│ 8884 │     0.184466 │
├──────┼──────────────┤
│ 6722 │     0.186895 │
├──────┼──────────────┤
│ 5160 │     0.187128 │
├──────┼──────────────┤
│ 3551 │     0.195014 │
├──────┼──────────────┤
│ 5330 │     0.196128 │
├──────┼──────────────┤
│ 4928 │     0.20922  │
├──────┼──────────────┤
│ 6942 │     0.21317  │
├──────┼──────────────┤
│ 2667 │     0.223906 │
╘══════╧══════════════╛


- For customer 3703

In [ ]:
print (tabulate(calc_10nn(3703,groceries),headers='firstrow',tablefmt='fancy_grid'))

╒══════╤══════════════╕
│   id │   similarity │
╞══════╪══════════════╡
│ 9942 │     0.120243 │
├──────┼──────────────┤
│ 1604 │     0.132479 │
├──────┼──────────────┤
│ 4838 │     0.136519 │
├──────┼──────────────┤
│ 3352 │     0.13841  │
├──────┼──────────────┤
│ 1837 │     0.144406 │
├──────┼──────────────┤
│ 3990 │     0.145034 │
├──────┼──────────────┤
│ 7194 │     0.14517  │
├──────┼──────────────┤
│ 7784 │     0.145247 │
├──────┼──────────────┤
│  374 │     0.145649 │
├──────┼──────────────┤
│ 6793 │     0.146853 │
╘══════╧══════════════╛


- For customer 4263

In [ ]:
print (tabulate(calc_10nn(4263,groceries),headers='firstrow',tablefmt='fancy_grid'))

╒══════╤══════════════╕
│   id │   similarity │
╞══════╪══════════════╡
│ 9536 │     0.119031 │
├──────┼──────────────┤
│ 4990 │     0.138546 │
├──────┼──────────────┤
│ 9051 │     0.14321  │
├──────┼──────────────┤
│ 2195 │     0.143373 │
├──────┼──────────────┤
│ 5829 │     0.155875 │
├──────┼──────────────┤
│ 3822 │     0.159324 │
├──────┼──────────────┤
│ 6183 │     0.160658 │
├──────┼──────────────┤
│ 5427 │     0.16229  │
├──────┼──────────────┤
│ 1896 │     0.16826  │
├──────┼──────────────┤
│ 5755 │     0.174091 │
╘══════╧══════════════╛


- For customer 5300

In [ ]:
print (tabulate(calc_10nn(5300,groceries),headers='firstrow',tablefmt='fancy_grid'))

╒══════╤══════════════╕
│   id │   similarity │
╞══════╪══════════════╡
│ 8497 │     0.126418 │
├──────┼──────────────┤
│ 8982 │     0.127739 │
├──────┼──────────────┤
│ 8711 │     0.128348 │
├──────┼──────────────┤
│ 2110 │     0.13121  │
├──────┼──────────────┤
│ 7457 │     0.134633 │
├──────┼──────────────┤
│ 3533 │     0.135548 │
├──────┼──────────────┤
│ 3470 │     0.138125 │
├──────┼──────────────┤
│ 8068 │     0.138177 │
├──────┼──────────────┤
│ 1999 │     0.13885  │
├──────┼──────────────┤
│ 8905 │     0.14144  │
╘══════╧══════════════╛


- For customer 6129

In [ ]:
print (tabulate(calc_10nn(6129,groceries),headers='firstrow',tablefmt='fancy_grid'))

╒══════╤══════════════╕
│   id │   similarity │
╞══════╪══════════════╡
│ 1082 │     0.109677 │
├──────┼──────────────┤
│ 6303 │     0.12252  │
├──────┼──────────────┤
│ 2029 │     0.128594 │
├──────┼──────────────┤
│ 7563 │     0.129656 │
├──────┼──────────────┤
│ 4933 │     0.131443 │
├──────┼──────────────┤
│ 6387 │     0.133074 │
├──────┼──────────────┤
│ 7870 │     0.134732 │
├──────┼──────────────┤
│ 5680 │     0.135302 │
├──────┼──────────────┤
│ 7557 │     0.137212 │
├──────┼──────────────┤
│ 5301 │     0.138203 │
╘══════╧══════════════╛


- For customer 7800

In [ ]:
print (tabulate(calc_10nn(7800,groceries),headers='firstrow',tablefmt='fancy_grid'))

╒══════╤══════════════╕
│   id │   similarity │
╞══════╪══════════════╡
│ 2126 │     0.117923 │
├──────┼──────────────┤
│  186 │     0.12129  │
├──────┼──────────────┤
│ 7470 │     0.144911 │
├──────┼──────────────┤
│ 2342 │     0.161577 │
├──────┼──────────────┤
│ 9116 │     0.165899 │
├──────┼──────────────┤
│  673 │     0.175136 │
├──────┼──────────────┤
│ 8293 │     0.177596 │
├──────┼──────────────┤
│ 1251 │     0.181974 │
├──────┼──────────────┤
│ 8212 │     0.181974 │
├──────┼──────────────┤
│ 1847 │     0.188889 │
╘══════╧══════════════╛


- For customer 8555

In [ ]:
print (tabulate(calc_10nn(8555,groceries),headers='firstrow',tablefmt='fancy_grid'))

╒══════╤══════════════╕
│   id │   similarity │
╞══════╪══════════════╡
│ 1486 │     0.123051 │
├──────┼──────────────┤
│ 6092 │     0.12476  │
├──────┼──────────────┤
│ 8732 │     0.131494 │
├──────┼──────────────┤
│ 3012 │     0.1331   │
├──────┼──────────────┤
│ 6823 │     0.133307 │
├──────┼──────────────┤
│ 3320 │     0.138436 │
├──────┼──────────────┤
│ 2691 │     0.139938 │
├──────┼──────────────┤
│ 3894 │     0.141777 │
├──────┼──────────────┤
│ 9336 │     0.141777 │
├──────┼──────────────┤
│ 5140 │     0.142547 │
╘══════╧══════════════╛


##  Customer rating prediction

We predict the rating of a customer with 2 methods : 
- The first is by calculating the average rating : we created a function that given a customer id and the dataset it calls the previous function in order to calculate the 10 most similar customers. The diference in the calling is that it first removes the column Customer_Rating from the dataset , since this is the one we want to predict. Note that as we noted earlier the function that calculates similarities will not take into account column Customer_Rating when the dataset is not full;having 10 columns. In this case the dataset has 9. After this the function sums the rating of the 10 most similar customers and returns the average rounded to the closest integer.
- The second is by calculating the weigthed average rating : we created a function that operates as the previous but it now calculates the weigthed average rounded to the closest integer.

In [ ]:
def pred_rating_avg(id,groceries):
    groceries_pred = groceries.drop(['Customer_Rating'], axis=1).copy()
    #first find the 10nn most similar customers
    sim_cust = calc_10nn(id,groceries_pred)
    rating = 0
    for list in sim_cust:
        #exclude the first row which is the header
        if list[0] != 'id':
            customer = list[0]
            #calculate the rating sum of top 10 most similar 
            rating = rating + groceries.iloc[customer].Customer_Rating
    #return the average
    return int(round(rating/10))

In [ ]:
def pred_rating_weight(id,groceries):
    groceries_pred = groceries.drop(['Customer_Rating'], axis=1).copy()
    #first find the 10nn most similar customers
    sim_cust = calc_10nn(id,groceries_pred)
    sum_similarities = 0
    weight_rating = 0
    for list in sim_cust:
        #exclude the first row which is the header
        if list[0] != 'id':
            customer = list[0]
            similarity = list[1]
            #calculate the sum of similarities
            sum_similarities = sum_similarities + similarity
            #calculate the multiplication of similarity score and customer rating
            sim_rank = similarity*groceries.iloc[customer].Customer_Rating
            weight_rating = weight_rating + sim_rank
    #return the weight average
    return int(round(weight_rating/sum_similarities))

Fo example customer 1 has 3 as his rating.

In [ ]:
groceries.iloc[1].Customer_Rating

3

If we predict his rating using the average rating of his/her 10 most similar customers we get 4 (if we do not round them in the closest integer the result is 3,5).

In [ ]:
pred_rating_avg(1,groceries)

4

Using the weighted average the result is also 4 (3,55 without rounding).

In [ ]:
pred_rating_weight(1,groceries)

4

To test our predictions we will calculate the Mean Prediction Error that measures the expected distance between what the predictor predicts for a specific value and what the true value is. In our case we will calculate the mean prediction error for the 50 first records.

For that purpose we created a function that takes as input the number of customers for which we want to calculate the Mean Prediction Error and the dataset. Then for all that customers it predicts his/her rating using both methods and finds the sum of error by substracting the actual value from the predicted one. Then it returns the error divided by the number of the input customers.

In [ ]:
def mpe(n,groceries):
    sum_error_avg = 0
    sum_error_weight = 0
    for i in range(n):
        #find the average prediction
        rating_avg = pred_rating_avg(i,groceries)
        #find the weighted average prediction
        rating_weight = pred_rating_weight(i,groceries)
        #calculate error for average prediction
        error_avg = abs(rating_avg - groceries.iloc[i].Customer_Rating )
        #calculate error for weighted average prediction
        error_weight = abs(rating_avg - groceries.iloc[i].Customer_Rating )
        #sum the results
        sum_error_avg = sum_error_avg + error_avg 
        sum_error_weight = sum_error_weight + error_weight
        
    #return the mean prediction error for both methods
    return sum_error_avg/n , sum_error_weight/n


We call the function for the first 50 customers.

In [ ]:
mpe = mpe(50,groceries)

In [ ]:
print('Mean Prediction Error for average prediction :  ' ,mpe[0])
print('Mean Prediction Error for weighted average prediction : ' ,mpe[1])

Mean Prediction Error for average prediction :   0.82
Mean Prediction Error for weighted average prediction :  0.82


We can see that the Mean Prediction Error for both methods is 0.82. This means that our predictions are not excellent  but neither too bad as the maximum error value is 4 (for example worst case scenario is the real value to be 5 and we predict 1). So, 0.82 means that we are almost certain that each result will not be further than one rating category. The error is the same in both methods because we round them to the closest integer. If we did not rounded them then the errors would be diferent, althought close to each other (as depicted before with the example with customer 1).